dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -221,249 +253,282 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
 
         try
         {
-            auto lock{ s_activatorInfoLock.lock_exclusive() };
-            if (s_pushTriggerRegistration)
+            Unregister();
+
+            auto lock{ m_lock.lock_exclusive() };
+            if (PushNotificationHelpers::IsPackagedAppScenario())
             {
-                s_pushTriggerRegistration.Unregister(true);
-                s_pushTriggerRegistration = nullptr;
-            }
+                THROW_HR_IF_NULL_MSG(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), m_pushTriggerRegistration, "PushTrigger not registered.");
 
-            s_comActivatorRegistration.reset();
+                m_pushTriggerRegistration.Unregister(true);
+                m_pushTriggerRegistration = nullptr;
+            }
 
-            if (s_protocolRegistration)
+            if (!AppModel::Identity::IsPackagedProcess())
             {
-                auto coInitialize = wil::CoInitializeEx();
+                THROW_HR_IF_MSG(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), !m_protocolRegistration, "Protocol Activator not registered");
+
+                auto coInitialize{ wil::CoInitializeEx() };
 
                 auto notificationPlatform{ PushNotificationHelpers::GetNotificationPlatform() };
 
                 wil::unique_cotaskmem_string processName;
                 THROW_IF_FAILED(GetCurrentProcessPath(processName));
                 LOG_IF_FAILED(notificationPlatform->UnregisterLongRunningActivator(processName.get()));
 
-                s_protocolRegistration = false;
+                m_protocolRegistration = false;
             }
         }
-        catch(...)
+        catch (...)
         {
-            PushNotificationTelemetry::ActivatorUnregisteredByApi(wil::ResultFromCaughtException(),
-                PushNotificationRegistrationActivators::PushTrigger | PushNotificationRegistrationActivators::ComActivator);
+            PushNotificationTelemetry::ActivatorUnregisteredByApi(wil::ResultFromCaughtException());
             throw;
         }
-        PushNotificationTelemetry::ActivatorUnregisteredByApi(S_OK, PushNotificationRegistrationActivators::PushTrigger | PushNotificationRegistrationActivators::ComActivator);
+        PushNotificationTelemetry::ActivatorUnregisteredByApi(S_OK);
     }
 
-    bool PushNotificationManager::IsActivatorSupported(PushNotificationRegistrationActivators const& activators)
+    winrt::event_token PushNotificationManager::PushReceived(TypedEventHandler<winrt::Microsoft::Windows::PushNotifications::PushNotificationManager, winrt::Microsoft::Windows::PushNotifications::PushNotificationReceivedEventArgs> handler)
     {
-        THROW_HR_IF(E_NOTIMPL, !::Microsoft::Windows::PushNotifications::Feature_PushNotifications::IsEnabled());
+        {
+            auto lock{ m_lock.lock_shared() };
+            THROW_HR_IF_MSG(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), m_comActivatorRegistration, "Must register event handlers before calling Register().");
```

Comment:
What about protocol? Why didn't we check that?

Reply:
@pmpurifoy

Reply:
Unpackaged scenarios will load their background payload in the command line which wouldn't conflict with the foreground scenario. BUT for consistency with packaged scenarios we should add the check.